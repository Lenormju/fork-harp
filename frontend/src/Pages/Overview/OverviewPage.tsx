import { Page } from "Components/Page"
import { OnQuerySuccess } from "Components/Utilities/OnQuerySuccess"
import { useOverviewDataQuery } from "Domain/Overview"
import { useSystemSettingsQuery } from "Domain/System"

import { TransactionsOverviewChart } from "./Components/OverviewChart/TransactionsOverviewChart"
import { TransactionsOverview } from "./OverviewCharts"

export const OverviewPage = () => {
  const query = useOverviewDataQuery()
  const settingsQuery = useSystemSettingsQuery()
  interface ProxyData {
    endpoints?: { name: string; port: number; url: string; description?: string }[]
  }

  return (
    <Page title="Overview" description="Useful insights">
      <OnQuerySuccess query={query}>
        {(query) => {
          return <TransactionsOverviewChart data={query.data} title="Transactions Overview" className="mb-10" />
        }}
      </OnQuerySuccess>
      <OnQuerySuccess query={settingsQuery}>
        {(query) => {
          const proxyData = query.data.proxy as ProxyData
          const endpointsNames = proxyData.endpoints?.map((endpoint: { name: string }) => endpoint.name)
          return (
            <div className="grid grid-cols-2 gap-4">
              {endpointsNames &&
                endpointsNames?.length > 1 &&
                endpointsNames.map((endpoint: string, index: number) => (
                  <TransactionsOverview key={index} endpoint={endpoint} title={endpoint} className="border p-2" />
                ))}
            </div>
          )
        }}
      </OnQuerySuccess>
    </Page>
  )
}
