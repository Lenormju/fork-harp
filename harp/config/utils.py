import importlib.util
from typing import TYPE_CHECKING, Type

from .defaults import DEFAULT_NAMESPACES

if TYPE_CHECKING:
    from .applications import Application


def get_application_class_name(name):
    return "".join(map(lambda x: x.title().replace("_", ""), name.rsplit(".", 1)[-1:])) + "Application"


def resolve_application_name(spec):
    if "." not in spec:
        for namespace in DEFAULT_NAMESPACES:
            _candidate = ".".join((namespace, spec))
            if importlib.util.find_spec(_candidate):
                return _candidate
    if importlib.util.find_spec(spec):
        return spec

    raise ModuleNotFoundError(f"No application named {spec}.")


#: Cache for application types
_application_types = {}


def get_application_type(name: str) -> Type["Application"]:
    """
    Returns the application class for the given application name.

    :param name:
    :return:
    """
    if name not in _application_types:
        application_spec = importlib.util.find_spec(name)
        if not application_spec:
            raise ValueError(f'Unable to find application "{name}".')

        try:
            application_module = __import__(".".join((application_spec.name, "__app__")), fromlist=["*"])
        except ModuleNotFoundError as exc:
            raise ModuleNotFoundError(
                f'A python package for application "{name}" was found but it is not a valid harp application. '
                'Did you forget to add an "__app__.py"?'
            ) from exc

        application_class_name = get_application_class_name(name)
        if not hasattr(application_module, application_class_name):
            raise AttributeError(f'Application module for {name} does not contain a "{application_class_name}" class.')

        _application_types[application_spec.name] = getattr(application_module, application_class_name)
        _application_types[application_spec.name].name = application_spec.name

    if name not in _application_types:
        raise RuntimeError(f'Unable to load application "{name}", application class definition not found.')

    return _application_types[name]
