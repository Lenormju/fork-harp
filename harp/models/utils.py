def generate_transaction_id_ksuid():
    from ksuid import KsuidMs

    return str(KsuidMs())


def generate_transaction_id_nanoid():
    from nanoid import generate
    from nanoid.resources import alphabet

    return generate(alphabet=alphabet[2:])


def generate_transaction_id_ulid():
    from ulid import ULID

    return str(ULID())
