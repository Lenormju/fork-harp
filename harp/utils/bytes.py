def ensure_bytes(x):
    if isinstance(x, bytes):
        return x
    return bytes(x, "utf-8")
