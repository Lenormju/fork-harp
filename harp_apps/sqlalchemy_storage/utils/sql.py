from sqlalchemy import text


async def run_sql(engine, sql, *, autocommit=True):
    if isinstance(sql, str):
        sql = text(sql)
    async with engine.connect() as conn:
        result = await conn.execute(sql)
        if autocommit:
            await conn.commit()
    return result
