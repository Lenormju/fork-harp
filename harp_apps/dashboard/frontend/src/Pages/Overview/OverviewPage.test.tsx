import { it } from "vitest"

import { renderWithClient } from "tests/utils"

import { MemoryRouter, Route, Routes } from "react-router-dom"
import OverviewPage from "./OverviewPage.tsx"
import { waitForElementToBeRemoved } from "@testing-library/react"

it.skip("renders the title and data when the query is successful", async () => {
  const result = renderWithClient(
    <MemoryRouter initialEntries={["/"]}>
      <Routes>
        <Route path="/" element={<OverviewPage />} />
      </Routes>
    </MemoryRouter>,
  )

  await waitForElementToBeRemoved(() => result.getAllByText("Loading..."), { timeout: 10000 })

  await result.findByText("endpoint1")

  // todo: for this to work, we need to ignore the difference in _echarts_instance_ attribute, which is a sequence
  //  starting at the initial timestamp. That could be overcome by using vitest fake timers, but then the suspense
  //  implementation wont work anymore. One way would be custom renderers, but too much time has already been spent
  //  here, for too little value.
  //expect(result.container).toMatchSnapshot()
}, { timeout: 10000 })
