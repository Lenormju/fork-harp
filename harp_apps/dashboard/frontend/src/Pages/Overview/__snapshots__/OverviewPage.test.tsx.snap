// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the title and data when the query is successful 1`] = `
<div>
  <main>
    <div
      class="mt-4 flex"
    >
      <div
        class="flex flex-col"
      >
        <h1
          class="css-9lrcch"
        >
          Overview
        </h1>
      </div>
    </div>
    <div
      class="mt-4"
    >
      <div
        class="sm:-mx-6 lg:-mx-8"
      >
        <section
          class="inline-block min-w-full max-w-full py-2 align-middle sm:px-6 lg:px-8"
        >
          <h2
            class="css-13yts7e"
          >
            Summary
          </h2>
          <div
            class="grid grid-cols-3 gap-4 mb-8"
          >
            <div
              class="css-9dvziv"
            >
              <h3
                class="css-1hpno4e"
              >
                Performances (24h)
              </h3>
              <div
                class="flex self-center relative z-10"
              >
                <span
                  class="bg-teal-400 ring-1 ring-white css-1g68g52"
                  title="score: 100"
                >
                  A++
                </span>
              </div>
              <svg
                preserveAspectRatio="none"
                style="margin: -3px; position: absolute; bottom: 0px; left: 0px; z-index: 0;"
                viewBox="0 0 240 60"
              >
                <g>
                  <circle
                    cx="2"
                    cy="2.5544554455445545"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="12.26086956521739"
                    cy="11.980198019801982"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="22.52173913043478"
                    cy="6.990099009900991"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="32.78260869565217"
                    cy="9.762376237623762"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="43.04347826086956"
                    cy="9.207920792079209"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="53.30434782608695"
                    cy="5.326732673267327"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="63.565217391304344"
                    cy="9.207920792079209"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="73.82608695652173"
                    cy="9.762376237623762"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="84.08695652173913"
                    cy="6.990099009900991"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="94.34782608695652"
                    cy="11.980198019801982"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="104.6086956521739"
                    cy="2.5544554455445545"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="114.8695652173913"
                    cy="11.980198019801982"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="125.13043478260869"
                    cy="6.990099009900991"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="135.39130434782606"
                    cy="9.762376237623762"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="145.65217391304347"
                    cy="9.207920792079209"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="155.91304347826087"
                    cy="5.326732673267327"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="166.17391304347825"
                    cy="9.207920792079209"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="176.43478260869563"
                    cy="9.762376237623762"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="186.69565217391303"
                    cy="6.990099009900991"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="196.95652173913044"
                    cy="11.980198019801982"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="207.2173913043478"
                    cy="2.5544554455445545"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="217.4782608695652"
                    cy="11.980198019801982"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="227.7391304347826"
                    cy="6.990099009900991"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="238"
                    cy="9.762376237623762"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <polyline
                    points="2 2.5544554455445545 12.26086956521739 11.980198019801982 22.52173913043478 6.990099009900991 32.78260869565217 9.762376237623762 43.04347826086956 9.207920792079209 53.30434782608695 5.326732673267327 63.565217391304344 9.207920792079209 73.82608695652173 9.762376237623762 84.08695652173913 6.990099009900991 94.34782608695652 11.980198019801982 104.6086956521739 2.5544554455445545 114.8695652173913 11.980198019801982 125.13043478260869 6.990099009900991 135.39130434782606 9.762376237623762 145.65217391304347 9.207920792079209 155.91304347826087 5.326732673267327 166.17391304347825 9.207920792079209 176.43478260869563 9.762376237623762 186.69565217391303 6.990099009900991 196.95652173913044 11.980198019801982 207.2173913043478 2.5544554455445545 217.4782608695652 11.980198019801982 227.7391304347826 6.990099009900991 238 9.762376237623762 238 58 2 58 2 2.5544554455445545"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <polyline
                    points="2 2.5544554455445545 12.26086956521739 11.980198019801982 22.52173913043478 6.990099009900991 32.78260869565217 9.762376237623762 43.04347826086956 9.207920792079209 53.30434782608695 5.326732673267327 63.565217391304344 9.207920792079209 73.82608695652173 9.762376237623762 84.08695652173913 6.990099009900991 94.34782608695652 11.980198019801982 104.6086956521739 2.5544554455445545 114.8695652173913 11.980198019801982 125.13043478260869 6.990099009900991 135.39130434782606 9.762376237623762 145.65217391304347 9.207920792079209 155.91304347826087 5.326732673267327 166.17391304347825 9.207920792079209 176.43478260869563 9.762376237623762 186.69565217391303 6.990099009900991 196.95652173913044 11.980198019801982 207.2173913043478 2.5544554455445545 217.4782608695652 11.980198019801982 227.7391304347826 6.990099009900991 238 9.762376237623762"
                    style="stroke: #ADD8E6; stroke-width: 1; stroke-linejoin: round; stroke-linecap: round; fill: none;"
                  />
                </g>
                <g>
                  <circle
                    cx="238"
                    cy="9.762376237623762"
                    r="2"
                    style="fill: #ADD8E6;"
                  />
                </g>
              </svg>
            </div>
            <div
              class="css-9dvziv"
            >
              <h3
                class="css-1hpno4e"
              >
                Transactions (24h)
              </h3>
              <div
                class="flex flex-col self-center relative z-10 items-start"
              >
                <span
                  class="bg-white ring-1 ring-blue-200 css-14jvdk3"
                  color="black"
                >
                  56
                   
                  <span
                    class="font-light"
                  >
                    /
                    min
                  </span>
                  <div
                    class="text-xs font-normal text-gray-500 text-right"
                  >
                    over the last 24 hours
                  </div>
                </span>
              </div>
              <svg
                preserveAspectRatio="none"
                style="margin: -3px; position: absolute; bottom: 0px; left: 0px; z-index: 0;"
                viewBox="0 0 240 60"
              >
                <g>
                  <circle
                    cx="2"
                    cy="17.75"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="12.26086956521739"
                    cy="2.875"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="22.52173913043478"
                    cy="10.75"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="32.78260869565217"
                    cy="6.375"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="43.04347826086956"
                    cy="7.25"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="53.30434782608695"
                    cy="13.375"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="63.565217391304344"
                    cy="7.25"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="73.82608695652173"
                    cy="6.375"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="84.08695652173913"
                    cy="10.75"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="94.34782608695652"
                    cy="2.875"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="104.6086956521739"
                    cy="17.75"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="114.8695652173913"
                    cy="2.875"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="125.13043478260869"
                    cy="10.75"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="135.39130434782606"
                    cy="6.375"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="145.65217391304347"
                    cy="7.25"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="155.91304347826087"
                    cy="13.375"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="166.17391304347825"
                    cy="7.25"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="176.43478260869563"
                    cy="6.375"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="186.69565217391303"
                    cy="10.75"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="196.95652173913044"
                    cy="2.875"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="207.2173913043478"
                    cy="17.75"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="217.4782608695652"
                    cy="2.875"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="227.7391304347826"
                    cy="10.75"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <circle
                    cx="238"
                    cy="6.375"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <polyline
                    points="2 17.75 12.26086956521739 2.875 22.52173913043478 10.75 32.78260869565217 6.375 43.04347826086956 7.25 53.30434782608695 13.375 63.565217391304344 7.25 73.82608695652173 6.375 84.08695652173913 10.75 94.34782608695652 2.875 104.6086956521739 17.75 114.8695652173913 2.875 125.13043478260869 10.75 135.39130434782606 6.375 145.65217391304347 7.25 155.91304347826087 13.375 166.17391304347825 7.25 176.43478260869563 6.375 186.69565217391303 10.75 196.95652173913044 2.875 207.2173913043478 17.75 217.4782608695652 2.875 227.7391304347826 10.75 238 6.375 238 58 2 58 2 17.75"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #ADD8E6; pointer-events: auto;"
                  />
                  <polyline
                    points="2 17.75 12.26086956521739 2.875 22.52173913043478 10.75 32.78260869565217 6.375 43.04347826086956 7.25 53.30434782608695 13.375 63.565217391304344 7.25 73.82608695652173 6.375 84.08695652173913 10.75 94.34782608695652 2.875 104.6086956521739 17.75 114.8695652173913 2.875 125.13043478260869 10.75 135.39130434782606 6.375 145.65217391304347 7.25 155.91304347826087 13.375 166.17391304347825 7.25 176.43478260869563 6.375 186.69565217391303 10.75 196.95652173913044 2.875 207.2173913043478 17.75 217.4782608695652 2.875 227.7391304347826 10.75 238 6.375"
                    style="stroke: #ADD8E6; stroke-width: 1; stroke-linejoin: round; stroke-linecap: round; fill: none;"
                  />
                </g>
                <g>
                  <circle
                    cx="238"
                    cy="6.375"
                    r="2"
                    style="fill: #ADD8E6;"
                  />
                </g>
              </svg>
            </div>
            <div
              class="css-9dvziv"
            >
              <h3
                class="css-1hpno4e"
              >
                Errors (24h)
              </h3>
              <div
                class="flex flex-col self-center relative z-10 items-start"
              >
                <span
                  class="bg-white ring-1 ring-red-200 css-14jvdk3"
                  color="black"
                >
                  11
                   
                  <span
                    class="font-light"
                  >
                    /
                    min
                  </span>
                  <div
                    class="text-xs font-normal text-gray-500 text-right"
                  >
                    over the last 24 hours
                  </div>
                </span>
              </div>
              <svg
                preserveAspectRatio="none"
                style="margin: -3px; position: absolute; bottom: 0px; left: 0px; z-index: 0;"
                viewBox="0 0 240 60"
              >
                <g>
                  <circle
                    cx="2"
                    cy="35.6"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="12.26086956521739"
                    cy="9.466666666666667"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="22.52173913043478"
                    cy="5.733333333333333"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="32.78260869565217"
                    cy="24.4"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="43.04347826086956"
                    cy="28.133333333333333"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="53.30434782608695"
                    cy="16.933333333333334"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="63.565217391304344"
                    cy="28.133333333333333"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="73.82608695652173"
                    cy="24.4"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="84.08695652173913"
                    cy="5.733333333333333"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="94.34782608695652"
                    cy="9.466666666666667"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="104.6086956521739"
                    cy="35.6"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="114.8695652173913"
                    cy="9.466666666666667"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="125.13043478260869"
                    cy="5.733333333333333"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="135.39130434782606"
                    cy="24.4"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="145.65217391304347"
                    cy="28.133333333333333"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="155.91304347826087"
                    cy="16.933333333333334"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="166.17391304347825"
                    cy="28.133333333333333"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="176.43478260869563"
                    cy="24.4"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="186.69565217391303"
                    cy="5.733333333333333"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="196.95652173913044"
                    cy="9.466666666666667"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="207.2173913043478"
                    cy="35.6"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="217.4782608695652"
                    cy="9.466666666666667"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="227.7391304347826"
                    cy="5.733333333333333"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <circle
                    cx="238"
                    cy="24.4"
                    r="2"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <polyline
                    points="2 35.6 12.26086956521739 9.466666666666667 22.52173913043478 5.733333333333333 32.78260869565217 24.4 43.04347826086956 28.133333333333333 53.30434782608695 16.933333333333334 63.565217391304344 28.133333333333333 73.82608695652173 24.4 84.08695652173913 5.733333333333333 94.34782608695652 9.466666666666667 104.6086956521739 35.6 114.8695652173913 9.466666666666667 125.13043478260869 5.733333333333333 135.39130434782606 24.4 145.65217391304347 28.133333333333333 155.91304347826087 16.933333333333334 166.17391304347825 28.133333333333333 176.43478260869563 24.4 186.69565217391303 5.733333333333333 196.95652173913044 9.466666666666667 207.2173913043478 35.6 217.4782608695652 9.466666666666667 227.7391304347826 5.733333333333333 238 24.4 238 58 2 58 2 35.6"
                    style="stroke: none; stroke-width: 0; fill-opacity: .1; fill: #e6b3ad; pointer-events: auto;"
                  />
                  <polyline
                    points="2 35.6 12.26086956521739 9.466666666666667 22.52173913043478 5.733333333333333 32.78260869565217 24.4 43.04347826086956 28.133333333333333 53.30434782608695 16.933333333333334 63.565217391304344 28.133333333333333 73.82608695652173 24.4 84.08695652173913 5.733333333333333 94.34782608695652 9.466666666666667 104.6086956521739 35.6 114.8695652173913 9.466666666666667 125.13043478260869 5.733333333333333 135.39130434782606 24.4 145.65217391304347 28.133333333333333 155.91304347826087 16.933333333333334 166.17391304347825 28.133333333333333 176.43478260869563 24.4 186.69565217391303 5.733333333333333 196.95652173913044 9.466666666666667 207.2173913043478 35.6 217.4782608695652 9.466666666666667 227.7391304347826 5.733333333333333 238 24.4"
                    style="stroke: #e6b3ad; stroke-width: 1; stroke-linejoin: round; stroke-linecap: round; fill: none;"
                  />
                </g>
                <g>
                  <circle
                    cx="238"
                    cy="24.4"
                    r="2"
                    style="fill: #e6b3ad;"
                  />
                </g>
              </svg>
            </div>
          </div>
          <div
            class="flex justify-end my-2"
          >
            <h2
              class="grow css-13yts7e"
            >
              History
            </h2>
            <span
              class="isolate inline-flex rounded-md shadow-sm"
            >
              <button
                class="css-1gwafep"
                type="button"
              >
                1h
              </button>
              <button
                class="css-1gwafep"
                type="button"
              >
                24h
              </button>
              <button
                class="css-v7wx26"
                type="button"
              >
                7d
              </button>
              <button
                class="css-1gwafep"
                type="button"
              >
                1m
              </button>
              <button
                class="css-1gwafep"
                type="button"
              >
                1y
              </button>
            </span>
          </div>
          <div
            class="mb-4 css-14upriz"
          >
            <div>
              <h3
                class="css-1hpno4e"
              >
                Transactions
              </h3>
              <div
                style="display: flex; align-items: center;"
              >
                <div
                  class="flex flex-col items-center"
                >
                  <div
                    class="flex self-center"
                  >
                    <span
                      class="bg-green-500 css-1g68g52"
                      title="score: 95"
                    >
                      A
                    </span>
                  </div>
                  <div
                    class="grid grid-cols-2 text-xs text-left align-text-bottom items-center self-center ml-10 mt-10"
                  >
                    <span
                      class="font-bold"
                    >
                      Mean duration:
                    </span>
                    <span>
                      0.03
                       s
                    </span>
                    <span
                      class="font-bold"
                    >
                      Errors:
                    </span>
                    <span>
                      50
                      %
                    </span>
                  </div>
                </div>
                <div
                  class="recharts-responsive-container"
                  style="width: 90%; height: 300px; min-width: 0;"
                />
              </div>
            </div>
          </div>
          <div
            class="grid grid-cols-2 gap-4"
          >
            <div
              class="css-14upriz"
            >
              <div>
                <h3
                  class="css-1hpno4e"
                >
                  endpoint1
                </h3>
                <div
                  style="display: flex; align-items: center;"
                >
                  <div
                    class="flex flex-col items-center"
                  >
                    <div
                      class="flex self-center"
                    >
                      <span
                        class="bg-green-500 css-1g68g52"
                        title="score: 95"
                      >
                        A
                      </span>
                    </div>
                    <div
                      class="grid grid-cols-2 text-xs text-left align-text-bottom items-center self-center ml-10 mt-10"
                    >
                      <span
                        class="font-bold"
                      >
                        Mean duration:
                      </span>
                      <span>
                        0.03
                         s
                      </span>
                      <span
                        class="font-bold"
                      >
                        Errors:
                      </span>
                      <span>
                        50
                        %
                      </span>
                    </div>
                  </div>
                  <div
                    class="recharts-responsive-container"
                    style="width: 90%; height: 300px; min-width: 0;"
                  />
                </div>
              </div>
            </div>
            <div
              class="css-14upriz"
            >
              <div>
                <h3
                  class="css-1hpno4e"
                >
                  endpoint2
                </h3>
                <div
                  style="display: flex; align-items: center;"
                >
                  <div
                    class="flex flex-col items-center"
                  >
                    <div
                      class="flex self-center"
                    >
                      <span
                        class="bg-green-500 css-1g68g52"
                        title="score: 95"
                      >
                        A
                      </span>
                    </div>
                    <div
                      class="grid grid-cols-2 text-xs text-left align-text-bottom items-center self-center ml-10 mt-10"
                    >
                      <span
                        class="font-bold"
                      >
                        Mean duration:
                      </span>
                      <span>
                        0.03
                         s
                      </span>
                      <span
                        class="font-bold"
                      >
                        Errors:
                      </span>
                      <span>
                        50
                        %
                      </span>
                    </div>
                  </div>
                  <div
                    class="recharts-responsive-container"
                    style="width: 90%; height: 300px; min-width: 0;"
                  />
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </main>
</div>
`;
